### PR TITLE
Fix lots of warnings and clang-tidy suggestions

### PR DESCRIPTION
**Pull request by jackgerrits**
Changes include:
- Change from deprecated c .h headers to <c...>headers
- Remove several redundant object copies in function parameters
- Remove unused includes
- Change loops to range for loops
- Use empty instead of size > 0
- Use string comparison operator rather than function
- Fix spelling errors
- Fix exception noexcept specification
- Mark virtual function overrides correctly

There are a lot of type mismatches in the codebase. I tackled a few but not many.

### CODE REVIEW

File: vowpalwabbit/active.cc [hunk 1]
```diff
@@ -117,9 +117,9 @@ void output_and_account_example(vw& all, active& a, example& ec)
     ai = query_decision(a, ec.confidence, (float)all.sd->weighted_unlabeled_examples);
 
   all.print(all.raw_prediction, ec.partial_prediction, -1, ec.tag);
-  for (size_t i = 0; i < all.final_prediction_sink.size(); i++)
+  for (int i : all.final_prediction_sink)
   {
-    int f = (int)all.final_prediction_sink[i];
+    int f = (int)i;
```
Comment (peterychang) [inline vowpalwabbit/active.cc]: is this necessary?

Comment (jackgerrits) [reply vowpalwabbit/active.cc]: Nope. Whoops auto fixed

File: vowpalwabbit/cb.cc [hunk 2]
```diff
@@ -59,9 +59,9 @@ char* bufcache_label(CB::label* ld, char* c)
 {
   *(size_t*)c = ld->costs.size();
   c += sizeof(size_t);
-  for (size_t i = 0; i < ld->costs.size(); i++)
+  for (auto cost : ld->costs)
```
Comment (peterychang) [inline vowpalwabbit/cb.cc]: const auto &?

File: vowpalwabbit/cb.cc [hunk 3]
```diff
@@ -87,10 +87,10 @@ void default_label(void* v)
 bool test_label(void* v)
 {
   CB::label* ld = (CB::label*)v;
-  if (ld->costs.size() == 0)
+  if (ld->costs.empty())
     return true;
-  for (size_t i = 0; i < ld->costs.size(); i++)
-    if (FLT_MAX != ld->costs[i].cost && ld->costs[i].probability > 0.)
+  for (auto & cost : ld->costs)
```
Comment (peterychang) [inline vowpalwabbit/cb.cc]: const auto &?

File: vowpalwabbit/cb.cc [hunk 4]
```diff
@@ -115,12 +115,12 @@ void parse_label(parser* p, shared_data*, void* v, v_array<substring>& words)
   ld->costs.clear();
   ld->weight = 1.0;
 
-  for (size_t i = 0; i < words.size(); i++)
+  for (auto word : words)
```
Comment (peterychang) [inline vowpalwabbit/cb.cc]: const auto &?

File: vowpalwabbit/cb_adf.cc [hunk 5]
```diff
@@ -154,7 +154,7 @@ void learn_SM(cb_adf& mydata, multi_learner& base, multi_ex& examples)
 
   // TODO: Check Marco's example that causes VW to report prob > 1.
 
-  for (uint32_t i = 0; i < mydata.prob_s.size(); i++)  // Scale example_wt by prob of chosen action
+  for (size_t i = 0; i < mydata.prob_s.size(); i++)  // Scale example_wt by prob of chosen action
```
Comment (peterychang) [inline vowpalwabbit/cb_adf.cc]: this and the for loop below can be range-based i think?

File: vowpalwabbit/cb_adf.cc [hunk 6]
```diff
@@ -235,15 +235,13 @@ void learn_MTR(cb_adf& mydata, multi_learner& base, multi_ex& examples)
...
 example* test_adf_sequence(multi_ex& ec_seq)
 {
-  if (ec_seq.size() == 0)
+  if (ec_seq.empty())
     THROW("cb_adf: At least one action must be provided for an example to be valid.");
 
   uint32_t count = 0;
   example* ret = nullptr;
-  for (size_t k = 0; k < ec_seq.size(); k++)
+  for (auto ec : ec_seq)
```
Comment (peterychang) [inline vowpalwabbit/cb_adf.cc]: auto *ec (to make it obvious its a pointer)

Comment (peterychang) [reply vowpalwabbit/cb_adf.cc]: also, const if possible?

File: vowpalwabbit/cb_adf.cc [hunk 7]
```diff
@@ -341,7 +338,7 @@ bool update_statistics(vw& all, cb_adf& c, example& ec, multi_ex* ec_seq)
     labeled_example = false;
 
   bool holdout_example = labeled_example;
-  for (size_t i = 0; i < ec_seq->size(); i++) holdout_example &= (*ec_seq)[i]->test_only;
+  for (auto & i : *ec_seq) holdout_example &= i->test_only;
```
Comment (peterychang) [inline vowpalwabbit/cb_adf.cc]: const auto &?

File: vowpalwabbit/cbify.cc [hunk 8]
```diff
@@ -322,13 +319,13 @@ void output_example(vw& all, example& ec, bool& hit_loss, multi_ex* ec_seq)
 
   if (!COST_SENSITIVE::cs_label.test_label(&ec.l))
   {
-    for (size_t j = 0; j < costs.size(); j++)
+    for (auto & cost : costs)
```
Comment (peterychang) [inline vowpalwabbit/cbify.cc]: const?

File: vowpalwabbit/comp_io.h [hunk 9]
```diff
@@ -21,19 +21,19 @@ class comp_io_buf : public io_buf
  public:
   std::vector<gzFile> gz_files;
 
-  virtual int open_file(const char* name, bool stdin_off, int flag = READ);
```
Comment (peterychang) [inline vowpalwabbit/comp_io.h]: are we sure these shouldn't be virtual?

File: vowpalwabbit/csoaa.cc [hunk 10]
```diff
@@ -171,20 +171,20 @@ struct ldf
...
 {
-  if (ec.indices.size() < 1)
+  if (ec.indices.empty())
     return false;
   if (ec.indices[0] != 'l')
     return false;
   v_array<COST_SENSITIVE::wclass> costs = ec.l.cs.costs;
-  for (size_t j = 0; j < costs.size(); j++)
-    if ((costs[j].class_index != 0) || (costs[j].x <= 0.))
+  for (auto & cost : costs)
```
Comment (peterychang) [inline vowpalwabbit/csoaa.cc]: const

File: vowpalwabbit/csoaa.cc [hunk 11]
```diff
@@ -582,13 +581,13 @@ void output_example(vw& all, example& ec, bool& hit_loss, multi_ex* ec_seq, ldf&
 
   if (!COST_SENSITIVE::cs_label.test_label(&ec.l))
   {
-    for (size_t j = 0; j < costs.size(); j++)
+    for (auto & cost : costs)
```
Comment (peterychang) [inline vowpalwabbit/csoaa.cc]: const

File: vowpalwabbit/csoaa.cc [hunk 12]
```diff
@@ -746,9 +745,9 @@ void inline process_label(ldf& data, example* ec)
 {
   auto new_fs = ec->feature_space[ec->indices[0]];
   auto& costs = ec->l.cs.costs;
-  for (size_t j = 0; j < costs.size(); j++)
+  for (auto & cost : costs)
```
Comment (peterychang) [inline vowpalwabbit/csoaa.cc]: const

File: vowpalwabbit/ect.cc [hunk 13]
```diff
@@ -88,19 +86,19 @@ size_t final_depth(size_t eliminations)
 
 bool not_empty(v_array<v_array<uint32_t>> tournaments)
 {
-  for (size_t i = 0; i < tournaments.size(); i++)
+  for (auto & tournament : tournaments)
```
Comment (peterychang) [inline vowpalwabbit/ect.cc]: const

File: vowpalwabbit/ect.cc [hunk 14]
```diff
@@ -88,19 +86,19 @@ size_t final_depth(size_t eliminations)
...
+    if (!tournament.empty())
       return true;
   }
   return false;
 }
 
 void print_level(v_array<v_array<uint32_t>> level)
 {
-  for (size_t t = 0; t < level.size(); t++)
+  for (auto & t : level)
```
Comment (peterychang) [inline vowpalwabbit/ect.cc]: const

File: vowpalwabbit/ezexample.h [hunk 15]
```diff
@@ -234,11 +234,11 @@ class ezexample
     quadratic_features_num = 0;
     quadratic_features_sqr = 0.;
 
-    for (std::vector<std::string>::iterator i = vw_ref->pairs.begin(); i != vw_ref->pairs.end(); i++)
+    for (auto & pair : vw_ref->pairs)
```
Comment (peterychang) [inline vowpalwabbit/ezexample.h]: const

File: vowpalwabbit/parser.cc [hunk 16]
```diff
@@ -769,7 +769,7 @@ example* read_example(vw& all, char* example_line)
   return ret;
 }
 
-example* read_example(vw& all, string example_line) { return read_example(all, (char*)example_line.c_str()); }
+example* read_example(vw& all, const string& example_line) { return read_example(all, (char*)example_line.c_str()); }
```
Comment (peterychang) [inline vowpalwabbit/parser.cc]: the const is a lie :(

Comment (peterychang) [reply vowpalwabbit/parser.cc]: maybe don't send refs for places we un-constify the strings

File: vowpalwabbit/vw.h [hunk 17]
```diff
@@ -64,12 +64,12 @@ struct primitive_feature_space  // just a helper definition.
 /* The simplest of two ways to create an example.  An example_line is the literal line in a VW-format datafile.
  */
 example* read_example(vw& all, char* example_line);
-example* read_example(vw& all, std::string example_line);
+example* read_example(vw& all, const std::string& example_line);
```
Comment (peterychang) [inline vowpalwabbit/vw.h]: don't send refs when we lie about constness

File: vowpalwabbit/comp_io.h (hunk 9 above)
Comment (jackgerrits) [reply vowpalwabbit/comp_io.h]: Yes, override is what we need to mark these correctly

**Pull request merged**